Add soft-delete documentation.

pop opened the pull request:
projects, activities, and times now have the 'deleted_at' and 'parent'
field for storing if they are an updated object ('parent' refers to the
original object) and if the object is deleted_at (which refers to if an
object has been archived or not).

There were also some miscellaneous formatting changes.

pop committed:
Added 'deleted_at' and 'parent' field to objects.

projects, activities, and times now have the 'deleted_at' and 'parent'
field for storing if they are an updated object ('parent' refers to the
original object) and if the object is deleted_at (which refers to if an
object has been archived or not).

There were also some miscellaneous formatting changes.

tschuy commented:
Yesss those docs are beautiful.

+1

MorganEPatch commented:
Looks very good to me. +1

Kennric commented on source/draft_api.rst:
```diff
@@ -223,7 +258,9 @@ Response body:
       "issue_uri":"https://github.com/osu-cass/whats-fresh-api/issues/56",
       "date_worked":null,
       "created_at":2014-09-18,
-      "updated_at":null
```
Is this the request body, or the response? If request, why are we sending updated, deleted, and parent? Shouldn't the back end generate these?

pop replied:
Oops, not sure how that snuck in.
Should we allow the user to also modify `updated_at` and `created_at`?

LyonesGamer replied:
That is the request body. Admittedly it's a tiny bit unclear.

Kennric replied:
Lets make it clear, and no, users should not be able to modify updated or created dates

MorganEPatch replied:
Trying to modify `created_at` would require going through and modifying the whole audit trail (`created_at` is constant in the trail for the "same" object). And I don't really see a benefit to modifying them. They're metadata, if they can be edited they on some level become meaningless.

pop committed:
Fixed request body containing update_* vars

Kennric commented:
I think adding a Request Body and Response Body sections to all the endpoints would be a good idea, it would be good to know both what to send and what to expect back.

pop commented:
@Kennric I will add that in this PR.

pop committed:
Addes 'response body' to all POST request examples

They should all be mostly accurate, they're just examples and hopefully
get the message across.

pop committed:
Increased resolution of deleted_at + created_at

Kennric commented on source/draft_api.rst:
```diff
@@ -209,21 +243,53 @@ Response body:
       "user": "example-2",
-      "project": "",
+      "project": "Ganeti Web Manager",
       "activities": ["gwm", "ganeti"],
```
These don't look like activities, they look like projects (and don't match activities in the response below)

pop replied:
I got a bit too copy-paste happy. Fixed it and pushing in the next commit.

athai commented on source/draft_api.rst:
```diff
@@ -69,7 +69,9 @@ GET Endpoints
-         "id": 1
+         "id": 1,
+         "deleted_at": null,
+         "parent": null
```
Don't we want an 'updated_at' field for projects/activities as well?

pop replied:
Yes we did, I'll add those now.

pop committed:
Fixed DELETE explanation and PATCH clarification.

pop committed:
Address feedback on PR #28

pop committed:
Added soft-delete relevant information.

Also updated previously existing content, including:
  - removed 'action' and 'version' parameters in favor of just
    'archived'
  - Added 'created_at' and 'updated_at' and 'deleted_at' parameters to
    all GET and POST response bodies (ensured they were there and
    accurate or the given examples)
fixes #30

Kennric commented on source/draft_api.rst:
```diff
@@ -172,47 +185,151 @@ GET Endpoints
+* user,
+* project,
+* activity, and
+* date range.
```
Bulleted lists don't need commas or 'and's

Kennric commented on source/draft_api.rst:
```diff
@@ -172,47 +185,151 @@ GET Endpoints
+* ``/times?activity=:activityslug``: Filters based on activity slug
+* ``/times?start=:date``: Filters to dates after and including the given date.
+* ``/times?end=:date``:  Filters to dates after and including the given date.
+* ``/times/?archived=:bool``: Filters on objects with ``deleted_at`` set to
```
The more I look at the term 'archived', the less I like it. Now that we are dealing with revisions, perhaps 'revisions' would be more apropos?

Also, this sentence says this filters on deleted_at that either is set or isn't set, which makes no sense. And the response actually has nothing to do with deleted_at being set, it just returns any extant revisions of a time.

Kennric commented on source/draft_api.rst:
```diff
@@ -261,32 +381,68 @@ Response body:
-      "project": "",
-      "activities": ["gwm", "ganeti"],
-      "notes":"",
+      "project": "Ganeti Web Manager",
```
I am not sure why this is a name string rather than a slug

Kennric commented:
I'd like a little more clarity on a few things, mostly dates.

We have a created_at, updated_at and deleted_at date all floating around, I'd like to propose that we use them like this:

created_at: original record belonging to this uuid created (carried by all records for uuid)
updated_at: object is revised, the create date of the new revision record
deleted_at: DELETE operation performed on this object (null if revised but not deleted)

pop committed:
Fixed feedback based on PR#28 feedback

pop committed:
Changed archived to revision, modified deleted_at.

Kennric commented on source/draft_api.rst:
```diff
@@ -172,47 +207,150 @@ GET Endpoints
+      "created_at": 2015-04-17,
+      "deleted_at": null,
+      "updated_at": null,
+      "parent":
```
parent, or parents? Since this field contains a list.

pop committed:
Merge branch 'develop' of github.com:osuosl/timesync into elijah/soft-delete

Conflicts:
	source/draft_api.rst

That was way too dicey for a docs merge...

pop committed:
s/parent/parents/g

Kennric commented:
+1

MaraJade commented:
+1
